boa/src/bytecompiler.rs

```diff
@@ -109,6 +120,49 @@ impl<'b> ByteCompiler<'b> {
         index
     }
 
+    #[inline]
+    fn get_or_insert_binding(&mut self, binding: BindingLocator) -> u32 {
+        if let Some(index) = self.bindings_map.get(&binding) {
+            return *index;
+        }
+
+        let index = self.code_block.bindings.len() as u32;
+        self.code_block.bindings.push(binding);
+        self.bindings_map.insert(binding, index);
+        index
+    }
+
+    #[inline]
+    fn emit_binding(&mut self, opcode: Opcode, name: Sym) {
+        match opcode {
+            Opcode::DefInitVar => {
+                let binding = if self.context.has_binding(name) {
+                    self.context.set_mutable_binding(name)
+                } else {
+                    self.context.initialize_mutable_binding(name, true)
+                };
+                let index = self.get_or_insert_binding(binding);
+                self.emit(opcode, &[index]);
+            }
+            Opcode::DefVar => {
+                let binding = self.context.initialize_mutable_binding(name, true);
+                let index = self.get_or_insert_binding(binding);
+                self.emit(opcode, &[index]);
+            }
+            Opcode::DefInitLet | Opcode::DefLet | Opcode::DefInitArg => {
+                let binding = self.context.initialize_mutable_binding(name, false);
+                let index = self.get_or_insert_binding(binding);
+                self.emit(opcode, &[index]);
+            }
+            Opcode::DefInitConst => {
+                let binding = self.context.initialize_immutable_binding(name);
+                let index = self.get_or_insert_binding(binding);
+                self.emit(opcode, &[index]);
+            }
+            _ => panic!("invalid opcode passed to emit_binding"),
```

Comment:
Could we add the opcode here to the panic? Also, adding the `#[track_caller]` to this function would help.

Reply:
Good catch. I added a specific enum, so that this match can be exhaustive.